### PR TITLE
Mutating admission controller initial version

### PR DESCRIPTION
**Pull request by Monkeyanator**
This commit serves as the initial substantial implementation for the mutating admission controller. Additional information on project architecture included in the README.

### CODE REVIEW

**Comment by dashpole** [comment]
I previously reviewed these changes while waiting for repo creation: https://github.com/Monkeyanator/mutating-trace-admission-controller-original/pull/1

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	docker push $(REGISTRY)/$(IMAGE):latest
+
+docker-tag-latest:
+	@echo 'Tagging image $(IMAGE) as latest...' 
+	docker tag $(IMAGE) $(REGISTRY)/$(IMAGE):latest
+
+# cluster commands
+cluster-up: deploy-certs deploy-config
+
+cluster-down: delete-certs delete-config
```
Comment (dashpole) [inline Makefile]: should we do this in reverse order of cluster-up?

File: main.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+			TLSConfig: &tls.Config{Certificates: []tls.Certificate{pair}},
+		},
+	}
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/mutate", whsvr.serve)
+	whsvr.server.Handler = mux
+
+	// TODO(Monkeyanator) pass configuration in through here
+	ConfigureTracing()
```
Comment (dashpole) [inline main.go:42]: TODOs should have a referenced issue associated with them

File: deploy/deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+spec:
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        app: trace-context-injector
+    spec:
+      containers:
+        - name: trace-context-injector
+          image: staging-k8s.gcr.io/trace-context-injector:latest
```
Comment (dashpole) [inline deploy/deployment.yaml]: I think this should just be `k8s.gcr.io`.  We push to staging, which is graduated to `k8s.gcr.io`

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	@echo 'Building image $(IMAGE)...'
+	docker build -t $(IMAGE) .
+
+docker-release: docker-build docker-publish
+
+docker-publish: docker-tag-latest
+	@echo 'Pushing image $(IMAGE) to $(REGISTRY)...'
+	docker push $(REGISTRY)/$(IMAGE):latest
+
+docker-tag-latest:
```
Comment (dashpole) [inline Makefile]: tagging and pushing to latest is probably a bit fine-grained.  I would prefer just having `make docker` build a docker image, and document the release process in a .md document.  The release documentation can be added later

File: webhook.go [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+)
+
+var (
+	runtimeScheme = runtime.NewScheme()
+	codecs        = serializer.NewCodecFactory(runtimeScheme)
+	deserializer  = codecs.UniversalDeserializer()
+)
+
+// TraceAnnotationKey is the annotation key the context should be injected at
+const TraceAnnotationKey string = "trace.kubernetes.io/context"
```
Comment (dashpole) [inline webhook.go:26]: @brancz is this a good name for the trace context annotation?

**Review by dashpole — COMMENTED** [review]
lgtm with a couple nits.

File: Makefile (hunk 4 above)
Comment (dashpole) [reply Makefile]: to expand on this, we should only add things to the makefile if we only expect a single variant.  I.E. we should avoid parameterized makefiles wherever possible

File: deploy/deployment.yaml (hunk 3 above)
Comment (Monkeyanator) [reply deploy/deployment.yaml]: done

File: main.go (hunk 2 above)
Comment (Monkeyanator) [reply main.go:42]: ack, removing TODO

File: Makefile (hunk 1 above)
Comment (Monkeyanator) [reply Makefile]: done

File: Makefile (hunk 4 above)
Comment (Monkeyanator) [reply Makefile]: done, will document release process in later PR

**Comment by dashpole** [comment]
Please open an issues for:

* The naming of the trace annotation
* The generation of the trace context 
  * should we start and not end a trace like we do now, or something different?
* Documenting the release process
* Improving the deployment process
  * Can we get to a point where users don't need to clone the repo + run scripts?
  * How will we support modifications to the provided configuration

/lgtm
/approve

**Comment by dashpole** [comment]
@Monkeyanator Can you squash your commits?

**Comment by Monkeyanator** [comment]
@dashpole  squashed

**Comment by coffeepac** [comment]
@dashpole you seem to not be a member of the `kubernetes-sigs` org yet.  This seems to be different from the `kubernetes` org.

**Comment by dashpole** [comment]
@coffeepac yeah, I am waiting for https://github.com/kubernetes/org/issues/286 to be resolved.  People are probably just busy at kubecon

**Pull request merged**